p2p/src/codec.rs

```diff
@@ -0,0 +1,137 @@
+use crate::core::ser::{BufReader, ProtocolVersion, Readable};
+use crate::msg::{MsgHeader, MsgHeaderWrapper};
+use crate::types::{AttachmentMeta, AttachmentUpdate, Error};
+use bytes::{BufMut, Bytes, BytesMut};
+use std::cmp::min;
+use std::io::Read;
+use std::net::TcpStream;
+use std::time::{Duration, Instant};
+use MsgHeaderWrapper::*;
+use State::*;
+
+const HEADER_IO_TIMEOUT: Duration = Duration::from_millis(2000);
+pub const BODY_IO_TIMEOUT: Duration = Duration::from_millis(60000);
+
+enum State {
+	None,
+	Header(MsgHeaderWrapper),
+	Attachment(usize, AttachmentMeta, Instant),
+}
+
+impl State {
+	fn take(&mut self) -> Self {
+		std::mem::replace(self, State::None)
+	}
+
+	fn is_none(&self) -> bool {
+		match self {
+			State::None => true,
+			_ => false,
+		}
+	}
+}
+
+pub enum Output {
```

Comment:
This jumped out to me as confusing due to `Output` being heavily used elsewhere in grin codebase.
Maybe we can rename this to something less generic?

Reply:
You're right, I will rename it. Naming structs isn't my strong suit, I don't really like the `Consume` and `Consumed` names either but couldn't come up with better ones